File: services/ai-detector/src/downloader.py
"""Fetching submission sources from the pre-signed URLs the job carries.

AID holds no MinIO credentials (D-80/D-85): the URL is the whole of its access,
signed against the internal endpoint for six hours. So this is a plain GET with
a retry budget, and its real job is classification — §2.5 asks services to
retry transient failures internally, at most three times, before reporting an
error.
"""

from __future__ import annotations

import time
from dataclasses import dataclass

import requests

from src.errors import PermanentError, TransientError

# D-34 caps a submission at 50KB; the ceiling here is generous headroom, meant
# to refuse something that is not a submission at all rather than to
# re-enforce the API's limit.
DEFAULT_MAX_BYTES = 1 << 20
DEFAULT_ATTEMPTS = 3
DEFAULT_BASE_DELAY_SECONDS = 0.5
DEFAULT_TIMEOUT_SECONDS = 30.0


@dataclass(frozen=True)
class Downloader:
    """Downloads sources. Immutable and safe to share."""

    max_bytes: int = DEFAULT_MAX_BYTES
    attempts: int = DEFAULT_ATTEMPTS
    base_delay_seconds: float = DEFAULT_BASE_DELAY_SECONDS
    timeout_seconds: float = DEFAULT_TIMEOUT_SECONDS
    # Injected in tests; None means the module-level `requests` API.
    http: object = None

    def download(self, url: str) -> str:
        """Returns the source as text, retrying transient failures.

        Raises PermanentError for a 4xx — an expired or wrong pre-signed URL
        answers the same way every time — or an oversized body, and
        TransientError once the attempt budget is spent.
        """
        delay = self.base_delay_seconds
        last: Exception | None = None

        for attempt in range(1, self.attempts + 1):
            try:
                return self._attempt(url)
            except PermanentError:
                raise
            except TransientError as error:
                last = error
                if attempt < self.attempts:
                    time.sleep(delay)
                    delay *= 2

        raise TransientError(f"downloading source after {self.attempts} attempts: {last}")

    def _attempt(self, url: str) -> str:
        client = self.http or requests

        try:
            response = client.get(url, timeout=self.timeout_seconds, stream=True)
        except requests.RequestException as error:
            raise TransientError(f"requesting source: {error}") from error

        with response:
            status = response.status_code
            if 400 <= status < 500:
                raise PermanentError(
                    f"source download answered {status} — the pre-signed URL is expired or wrong"
                )
            if status != 200:
                raise TransientError(f"source download answered {status}")

            # Streamed and capped: a body larger than any submission this
            # system stores is refused rather than read into memory.
            chunks: list[bytes] = []
            size = 0
            try:
                for chunk in response.iter_content(chunk_size=64 * 1024):
                    if not chunk:
                        continue
                    chunks.append(chunk)
                    size += len(chunk)
                    if size > self.max_bytes:
                        raise PermanentError(f"source exceeds the {self.max_bytes} byte cap")
            except requests.RequestException as error:
                raise TransientError(f"reading source body: {error}") from error

        # Student source is scored as text; a file that is not decodable as
        # UTF-8 is not source this system stored.
        try:
            return b"".join(chunks).decode("utf-8")
        except UnicodeDecodeError as error:
            raise PermanentError(f"source is not valid UTF-8: {error}") from error

File: services/plagiarism-checker/internal/downloader/downloader.go
// Package downloader fetches submission sources from the pre-signed URLs the
// job message carries.
//
// SIM holds no MinIO credentials (D-80/D-85): the URL is the whole of its
// access, signed against the internal endpoint and valid for six hours. So
// this is a plain HTTP GET with a retry budget, and its only real job is to
// classify what went wrong — §2.5 asks services to retry transient failures
// internally before reporting an error, and `shared/go/rmq` retries whatever
// escapes classified as Transient.
package downloader

import (
	"context"
	"fmt"
	"io"
	"net/http"
	"time"

	"github.com/magecode/shared/go/apperror"
)

// Config bounds one download. Zero values take the documented defaults.
type Config struct {
	// MaxBytes caps a single body. D-34 caps a submission at 50KB, so
	// anything past a generous multiple of that is not a submission this
	// system wrote.
	MaxBytes int64
	// Attempts is the total number of tries, not the number of retries.
	Attempts int
	// BaseDelay is the first backoff; each retry doubles it.
	BaseDelay time.Duration
	// Timeout bounds one attempt, not the whole budget.
	Timeout time.Duration
}

const (
	defaultMaxBytes  = 1 << 20
	defaultAttempts  = 3
	defaultBaseDelay = 500 * time.Millisecond
	defaultTimeout   = 30 * time.Second
)

func (c Config) withDefaults() Config {
	if c.MaxBytes <= 0 {
		c.MaxBytes = defaultMaxBytes
	}
	if c.Attempts <= 0 {
		c.Attempts = defaultAttempts
	}
	if c.BaseDelay <= 0 {
		c.BaseDelay = defaultBaseDelay
	}
	if c.Timeout <= 0 {
		c.Timeout = defaultTimeout
	}
	return c
}

// Client downloads submission sources. It is safe for concurrent use.
type Client struct {
	cfg  Config
	http *http.Client
}

// New returns a Client with cfg's defaults filled in.
func New(cfg Config) *Client {
	cfg = cfg.withDefaults()
	return &Client{cfg: cfg, http: &http.Client{Timeout: cfg.Timeout}}
}

// Download fetches url, retrying transient failures within the attempt budget.
//
// The returned error is always classified: Permanent for a 4xx (an expired or
// wrong URL reads the same on every attempt) or an oversized body, Transient
// for anything the network might do differently next time.
func (c *Client) Download(ctx context.Context, url string) ([]byte, error) {
	delay := c.cfg.BaseDelay
	var lastErr error

	for attempt := 1; attempt <= c.cfg.Attempts; attempt++ {
		if err := ctx.Err(); err != nil {
			return nil, apperror.Wrap(apperror.Transient, "download cancelled", err)
		}

		body, err := c.attempt(ctx, url)
		if err == nil {
			return body, nil
		}
		if apperror.IsPermanent(err) {
			return nil, err
		}
		lastErr = err

		if attempt < c.cfg.Attempts {
			select {
			case <-ctx.Done():
				return nil, apperror.Wrap(apperror.Transient, "download cancelled", ctx.Err())
			case <-time.After(delay):
			}
			delay *= 2
		}
	}

	return nil, apperror.Wrap(apperror.Transient,
		fmt.Sprintf("downloading source after %d attempts", c.cfg.Attempts), lastErr)
}

func (c *Client) attempt(ctx context.Context, url string) ([]byte, error) {
	request, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		// A URL the http package will not even build a request for came
		// from the message, and the message will not change.
		return nil, apperror.Wrap(apperror.Permanent, "building download request", err)
	}

	response, err := c.http.Do(request)
	if err != nil {
		return nil, apperror.Wrap(apperror.Transient, "requesting source", err)
	}
	defer func() { _ = response.Body.Close() }()

	if response.StatusCode >= http.StatusBadRequest && response.StatusCode < http.StatusInternalServerError {
		return nil, apperror.New(apperror.Permanent,
			fmt.Sprintf("source download answered %d %s", response.StatusCode, http.StatusText(response.StatusCode)))
	}
	if response.StatusCode != http.StatusOK {
		return nil, apperror.New(apperror.Transient,
			fmt.Sprintf("source download answered %d %s", response.StatusCode, http.StatusText(response.StatusCode)))
	}

	// One byte past the cap is read deliberately: it is how an oversized
	// body is told apart from one that exactly fills it.
	body, err := io.ReadAll(io.LimitReader(response.Body, c.cfg.MaxBytes+1))
	if err != nil {
		return nil, apperror.Wrap(apperror.Transient, "reading source body", err)
	}
	if int64(len(body)) > c.cfg.MaxBytes {
		return nil, apperror.New(apperror.Permanent,
			fmt.Sprintf("source exceeds the %d byte cap", c.cfg.MaxBytes))
	}

	return body, nil
}
